### ISSUE TITLE
Matrix and Array should be indexed like NumPy

### ISSUE THREAD

**Issue by Upabjojr** [sympy/sympy#15464]
In NumPy if you give only one index to a matrix, you get the row corresponding to that index. This is compatible with the list of list representations.

Matrices and arrays in SymPy only return one element, corresponding to the element of the flattened array.

Shall we change this and turn the behaviour into the one of NumPy?

**Comment by moorepants** [comment]
That sounds like a major backwards incompatible change on the face of it. i'd be -1 simply because of that. i have no problem with the Matrix API being slightly different from a NumPy array since they don't represent the same things. But when making decisions for new API designs, it seems following the highly popular "standard" NumPy has set is a good general practice. As for SymPy Arrays I'm not sure how widely used they are or how long they've been in our code base, depending on various factors it could be a reasonable decision to change the API. Some examples would help understand the issue.

**Comment by asmeurer** [comment]
I agree Array should. I think that was the original point, to work like NumPy. For Matrix, it probably doesn't make sense to change compatibility, since it already isn't like a NumPy array (for instance, it always has rank 2).

**Comment by bjodah** [comment]
I agree with @Upabjojr that sympy's Matrix has a surprising syntax (I expected NumPy convention when I started using it). But I also agree that it's too late to change.

"Slightly" off-topic: why doesn't SymPy just adhere to semantic versioning, then we can accumulate a large number of breaking changes and release `sympy2`? (perhaps along with `sympy1`, and have `sympy` be an "alias" for latest sympy).

**Comment by Upabjojr** [comment]
Other issues arising from this choice:
```python
In [31]: torch.DoubleTensor(eye(3))
Out[31]: tensor([1., 0., 0., 0., 1., 0., 0., 0., 1.], dtype=torch.float64)
```
The library `torch` casts the SymPy matrix into a one-dimensional array. It does not happen if you have a list of lists:
```python
In [32]: torch.DoubleTensor(eye(3).tolist())
Out[32]: 
tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]], dtype=torch.float64)
```

**Comment by Upabjojr** [comment]
And it is correct with NumPy:
```python
In [40]: torch.DoubleTensor(np.eye(3))
Out[40]: 
tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]], dtype=torch.float64)
```

**Comment by moorepants** [comment]
> why doesn't SymPy just adhere to semantic versioning

We should!

**Comment by sylee957** [comment]
It is a notable difference that SymPy's matrix have different length.

```python
import sympy
import numpy

array = [[1, 2], [3, 4]]
print(len(array))
print(len(sympy.Matrix(array)))
print(len(numpy.array(array)))
```
